File: app/src/main/java/com/shar2wy/moviesapp/models/reviewRepo/ReviewRemoteDataSource.java
package com.shar2wy.moviesapp.models.reviewRepo;

import android.content.Context;

import com.shar2wy.moviesapp.network.ApiClient;
import com.shar2wy.moviesapp.network.ApiService;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;

/**
 * Created by shar2wy
 * on 4/4/18.
 * Description: description goes here
 */
class ReviewRemoteDataSource {

    ApiService apiService;
    String apiKey = "38db4ee0b3b9fe63c95b9c835b73021f";

    public ReviewRemoteDataSource(Context context) {
        apiService = ApiClient.getClient(context).create(ApiService.class);
    }

    public Flowable<List<Review>> getReviews(int id) {
        return apiService.getMovieReviews(id, apiKey)
                .flatMap(reviewsResponse -> Flowable.just(reviewsResponse.getResults()))
                .flatMapIterable(reviews -> reviews)
                .map(review -> {
                    review.setMovieId(id);
                    return review;
                })
                .toList()
                .toFlowable()
                .subscribeOn(AndroidSchedulers.mainThread());
    }
}

File: app/src/main/java/com/shar2wy/moviesapp/models/reviewRepo/ReviewRemoteDataSource.kt
package com.shar2wy.moviesapp.models.reviewRepo

import android.content.Context
import com.shar2wy.moviesapp.network.ApiClient
import com.shar2wy.moviesapp.network.ApiService
import com.shar2wy.moviesapp.util.Constants.apiKey
import io.reactivex.Flowable

/**
 * Created by shar2wy
 * on 4/4/18.
 * Description: description goes here
 */
internal class ReviewRemoteDataSource(context: Context) {

    var apiService: ApiService = ApiClient.getInstance(context).getService(ApiService::class.java)

    fun getReviews(id: Int): Flowable<List<Review>> {
        return apiService.getMovieReviews(id, apiKey)
                .flatMap { reviewsResponse -> Flowable.just(reviewsResponse.results) }
                .flatMapIterable { reviews -> reviews }
                .map { review ->
                    review.movieId = id
                    review
                }
                .toList()
                .toFlowable()
    }
}
